### DISCUSSION TITLE
[OP Stack Deployment] Need help making L2 transactions gasless

### DISCUSSION THREAD

**Discussion by surhud004** [ethereum-optimism/developers#622] [Deploying the OP Stack] [answered]
### Did you check the documentation?

- [X] Yes

### Did you check for duplicate questions?

- [X] Yes

### Issue Description

Hey team, I need some dev support here.

I am running my own custom L2 chain using OP stack and I followed this tutorial -
https://docs.optimism.io/builders/chain-operators/tutorials/create-l2-rollup

I am running my custom L2 with Holesky L1 testnet locally. So far, everything is up and running smoothly where I can create and submit transactions on my custom L2 and they work successfully.

Now I want to modify it to set the gas price to zero on my L2 so that all transactions would be gasless on L2. More specifically, I want to set the L2 Execution Gas Fee to zero by setting the L2 Gas Price to zero. Is that possible? Can someone guide me on what fields or configurations should I modify to achieve this locally? Thank you in advance :D

### Logs

Logs screenshot -
* Top-left image: `op-geth`
* Bottom-left image: `op-node`
* Top-right image: `op-batcher`
* Bottom-right image: `op-proposer`

![image (3)-20241001-201002](https://github.com/user-attachments/assets/6a0451aa-f9ee-4e99-8b65-f759c0fcf9f9)

**Comment by sbvegan** [comment, marked as answer]
This one is tricky. The OP Stack uses the same EIP-1559 gas market as Ethereum. So you would need to modify the execution layer to do this on the L2. Here's a high-level doc on modifying the EL and something to get you started: https://docs.optimism.io/builders/chain-operators/hacks/execution

**Reply by sbvegan** [reply]
Alternatively you can look at the custom gas token implementation of the code and see if you can modify it so the CGT handles the L2 side and make that token free to mint and then the sequencer covers the L1 cost.

**Reply by surhud004** [reply]
Hey @sbvegan thank you for your quick response. I am looking into the alternative option you mentioned - to use a CGT. I wanted to know if I am able to set the gas price of the CGT to be exact zero and automate the minting process instead. 

I am happy to pay for the execution costs on L1 on behalf of the user. However, my goal is to improve the user interaction experience on my custom L2 network. Users should be able to connect to L2 and execute transactions directly without any additional steps involved such as the need of requesting the CGT from a faucet and/or enabling it in their wallet. They should just connect and execute transactions without managing the gas costs. Do you think this is achievable? Could you please guide me through this and the modifications I would need to make? Thanks! :D

**Reply by surhud004** [reply]
Hey @sbvegan any update on this? I was able to deploy the custom gas token and integrate with L2. However, I am still running into issues with handling L1 fee component costs via sequencer. It is currently being charged to the sender wallet so users still need to have some amount of gas token in their wallet for executing transactions. How can we configure the sequencer to cover these costs? Thanks in advance :D

**Comment by seromenho** [comment]
Hi everyone,

Achieving this would involve modifying the L2 execution layer to adjust the gas model or implementing a custom gas token mechanism where the sequencer covers L1 costs.

Before diving deeper, I wanted to confirm: is there already an existing "hack" or workaround for this that someone in the community might have implemented? Any pointers or shared experiences would be greatly appreciated!

Thanks in advance for your help!
